mijia/src/bluetooth/events.rs

```diff
@@ -57,96 +61,144 @@ pub enum CharacteristicEvent {
 }
 
 impl BluetoothEvent {
-    /// Return a `MatchRule` which will match all D-Bus messages which represent Bluetooth events.
-    pub(crate) fn match_rule() -> MatchRule<'static> {
+    /// Return a set of `MatchRule`s which will match all D-Bus messages which represent Bluetooth
+    /// events, possibly limited to those for a particular object (such as a device, service or
+    /// characteristic).
+    ///
+    /// Note that the match rules for a device will not match the device discovered event for that
+    /// device, as it is considered an event for the system rather than the device itself.
+    pub(crate) fn match_rules(object: Option<impl Into<Path<'static>>>) -> Vec<MatchRule<'static>> {
         // BusName validation just checks that the length and format is valid, so it should never
         // fail for a constant that we know is valid.
         let bus_name = "org.bluez".into();
-        PropertiesPropertiesChanged::match_rule(Some(&bus_name), None).static_clone()
+
+        let mut match_rules = vec![];
+
+        // If we aren't filtering to a single device or characteristic, then match ObjectManager
+        // signals so we can get events for new devices being discovered.
+        if object.is_none() {
+            let match_rule =
+                ObjectManagerInterfacesAdded::match_rule(Some(&bus_name), None).static_clone();
+            match_rules.push(match_rule);
+        }
```

Comment:
(optional) I was already a bit skeptical about match_rule() living inside BluetoothEvent (but it is not public api so I didn't say anything). I think I would split this into a couple of functions in the same module as the caller (just `filtered_event_stream()` currently), and make the caller responsible for calling this as a separate function (but it is not public api so I'm also okay with it as it is).

Reply:
I quite like having this logic in the same module as `message_to_events`, because it kind of parallels it. I think of this module as having as much as possible of the logic for dealing with events without having anything async or actually talking to D-Bus, which means it can easily be tested.